File: 0022-generate-parentheses/0022-generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    let answ = []
    backtrack("", 0, 0)
    
    function backtrack(curr, openingCount, closingCount){
        if(curr.length == 2*n){
            answ.push(curr)
            return
        }
        if(openingCount == n){
            backtrack(curr + ")", openingCount, closingCount + 1)
        }
        if(openingCount != n){
            backtrack(curr + "(", openingCount + 1, closingCount)
            if(closingCount < openingCount){
                backtrack(curr + ")", openingCount, closingCount + 1)
            }
        }
    }
    return answ;
};

File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> answ;
        solve(n, "", 0, 0, answ);
        return answ;
    }
    void solve(int n, string curr, int leftCount, int rightCount, vector<string>& s){
        if(curr.size() == n * 2){
            s.push_back(curr);            
            return;
        }

        // (
        if(leftCount < n){
            solve(n, curr + '(', leftCount + 1, rightCount, s);
        }
        // )        
        if(leftCount > rightCount){
            solve(n, curr + ')', leftCount, rightCount+1, s);
        }
    }
};